RecoJets/JetProducers/plugins/HFJetShowerShape.cc

```diff
@@ -0,0 +1,229 @@
+        double dphi = std::abs(deltaPhi(icand->phi(), jet.phi()));
+        double pt_PUsub = icand->pt() - puoffset;
+
+        //This is simply the size of the central eta strip and the adjacent strips
+        if (pt_PUsub >= stripPtThreshold_) {
+          if (dphi <= hfTowerPhiWidth_ * 0.5)
+            size_CentralEtaStrip++;
+          else if (dphi <= hfTowerPhiWidth_ * 1.5)
+            size_AdjacentEtaStrips++;
+        }
+
+        //Now computing sigmaEtaEta/PhiPhi
+        if (pt_PUsub >= widthPtThreshold_ && sumptPFcands > 0) {
+          double weight = pt_PUsub / sumptPFcands;
+          sigmaEtaEtaSq += deta * deta * weight;
+          sigmaPhiPhiSq += dphi * dphi * weight;
+          sumweightsPFcands += weight;
+        }
+      }
+
+      v_size_CentralEtaStrip.push_back(size_CentralEtaStrip);
+      v_size_AdjacentEtaStrips.push_back(size_AdjacentEtaStrips);
+
+      if (sumweightsPFcands > 0 && sigmaEtaEtaSq > 0 && sigmaPhiPhiSq > 0) {
+        v_sigmaEtaEta.push_back(sqrt(sigmaEtaEtaSq / sumweightsPFcands));
+        v_sigmaPhiPhi.push_back(sqrt(sigmaPhiPhiSq / sumweightsPFcands));
+      } else {
+        v_sigmaEtaEta.push_back(-1.);
+        v_sigmaPhiPhi.push_back(-1.);
+      }
+
+    }  //End loop over jets
+  }
+
+  putInEvent("sigmaEtaEta", theJets, v_sigmaEtaEta, iEvent);
+  putInEvent("sigmaPhiPhi", theJets, v_sigmaPhiPhi, iEvent);
+  putInEvent("centralEtaStripSize", theJets, v_size_CentralEtaStrip, iEvent);
+  putInEvent("adjacentEtaStripsSize", theJets, v_size_AdjacentEtaStrips, iEvent);
+}
+
+/// Function to put product into event
+template <typename T>
+void HFJetShowerShape::putInEvent(const std::string& name,
+                                  const edm::Handle<edm::View<reco::Jet>>& jets,
+                                  std::vector<T> product,
+                                  edm::Event& iEvent) const {
+  auto out = std::make_unique<edm::ValueMap<T>>();
+  typename edm::ValueMap<T>::Filler filler(*out);
+  filler.insert(jets, product.begin(), product.end());
+  filler.fill();
+  iEvent.put(std::move(out), name);
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void HFJetShowerShape::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("theJets", edm::InputTag("ak4PFJetsCHS"));
+  desc.add<edm::InputTag>("theVertices", edm::InputTag("offlinePrimaryVerticesWithBS"));
+  desc.add<double>("jetPtThreshold", 25.);
+  desc.add<double>("jetEtaThreshold", 2.9);
+  desc.add<double>("hfTowerEtaWidth", 0.175);
+  desc.add<double>("hfTowerPhiWidth", 0.175);
+  desc.add<double>("vertexRecoEffcy", 0.7);
+  desc.add<double>("offsetPerPU", 0.4);
+  desc.add<double>("jetReferenceRadius", 0.4);
+  desc.add<double>("stripPtThreshold", 10.);
+  desc.add<double>("widthPtThreshold", 3.);
+  descriptions.add("hfJetShowerShape", desc);
```

Comment:
If you rely on the config automatically produced by fillDescriptions, the name of both the file and the module have now the lower case "hf" ("hfJetShowerShape_cfi" and "hfJetShowerShape" respectively).

You should either rename them everywhere they are called in your configs, or rename as "HFJetShowerShape" here in the `descriptions.add` method.

Alternatively, you maintain the HFJetShowerShape_cfi.py file with the explicit config: in that file you can clone HFJetShowerShape from hfJetShowerShape_cfi.py. In that way you'd have an explicit config in which you can modify any of the thresholds set in case you decide to further tune them in the future. In case those threshold parameters are not supposed to last forever I would prefere this second option, which let you some handle for future modifications in the configuration without having to touch the C++ code.

Reply:
Yes I saw that yesterday evening... I think I forgot to scram clean when doing the tests before updating the PR, which is why they succeeded locally. Sorry for that. 
I've corrected this (following) and I'm running the tests again. But now that you mention it, it's probably better to keep a separate cfi file (I do expect some retuning of these parameters indeed...).

Reply:
I don't see a need for a separate file if only one instance of the module is expected.
It would even be rather confusing to have parameters inconsistently defined.
What is the plan? Will there be multiple instances of the same module?

Reply:
We expect only one instance indeed. Some of the current parameters might be retuned at some point but then it might be better to update the default ones even if it implies an update of the producer class. Ok I'll switch back to Slava's initial suggestion and drop the separate cfi file.

Reply:
> I don't see a need for a separate file if only one instance of the module is expected.
> It would even be rather confusing to have parameters inconsistently defined.
> What is the plan? Will there be multiple instances of the same module?

I never suggested to make two different instances of the same module, but rather to clone the default one into the one actually used in the config. So that you don't have to recompile the c++ code if you want to modify one of those parameters, which is exactly the reason why one normally makes those parameters configurable.

In any case, I won't insist about the implementation: let proceed as you and @slava77 believe is more convenient

Reply:
I guess we can discuss separately from this PR. In my gedanken experiment, I'm getting conflicts with one definition rule.
A default configuration, which is not equal to the one defined separately in pyhon configs implies multiple instances. 
Your proposal for a case where only one instance is expected implies either diverging configurations or the parameters that could be modified would have to be inserted in the fillDescriptions without a default (a required parameter).

Reply:
I guess a good legitimate reason is the era modifier support.

Reply:
Indeed this came to my mind too but I don't expect era dependent tuning to happen, at least not in the short future (also because HF didn't change that much during Run 2 and is not expect to change much in Run 3 from what I understand). 

Anyways I went ahead and fixed the misspelled calls to the producer.

Reply:
as @perrotta pointed out in our chat, this PR shows that multiple instances of this module are needed (one for running on AOD, one for running on miniAOD, and one for running in a nanoAOD job).

This is a pattern where the more appropriate solution is to make a default configuration in the fillDescriptions as described 
```C++
   descriptions.add("hfJetShowerShapeDefault", desc);
```
which can then be cloned in the actual relevant instances.
This will also "naturally" lead to making a hfJetShowerShape_cfi.py file, which can potentially have tuning for other eras (even though Run-2 is not that different from Run-2, the phase-2 setup likely is fairly different).

It looks like a in all these cases theJets and theVertices will need to be changed.
To make this a requirement, these parameters in the fillDescriptions can be left without a default
```C++
  desc.add<edm::InputTag>("jets");
  desc.add<edm::InputTag>("vertices");
```
(I also dropped the not obviously necessary prefix `the`)

On a related topic, would it make more sense to run this module on `offlineSlimmedPrimaryVertices`  instead of `offlinePrimaryVerticesWithBS` for more consistency between the multiple variants (on-AOD, on-mini)?